Libplanet.Explorer/Indexing/BlockChainIndexBase.cs

```diff
@@ -0,0 +1,468 @@
+        GetBlockHashesByOffsetAsync(int? offset, int? limit, bool desc, Address? miner);
+
+    /// <inheritdoc />
+    public abstract IEnumerable<TxId>
+        GetSignedTxIdsByAddress(Address signer, int? offset, int? limit, bool desc);
+
+    /// <inheritdoc />
+    public abstract IAsyncEnumerable<TxId>
+        GetSignedTxIdsByAddressAsync(Address signer, int? offset, int? limit, bool desc);
+
+    /// <inheritdoc />
+    public abstract IEnumerable<TxId>
+        GetInvolvedTxIdsByAddress(Address address, int? offset, int? limit, bool desc);
+
+    /// <inheritdoc />
+    public abstract IAsyncEnumerable<TxId>
+        GetInvolvedTxIdsByAddressAsync(Address address, int? offset, int? limit, bool desc);
+
+    /// <inheritdoc />
+    public abstract long? GetLastNonceByAddress(Address address);
+
+    /// <inheritdoc />
+    public abstract Task<long?> GetLastNonceByAddressAsync(Address address);
+
+    /// <inheritdoc />
+    public BlockHash GetContainedBlockHashByTxId(TxId txId) =>
+        TryGetContainedBlockHashById(txId, out var containedBlock)
+            ? containedBlock
+            : throw new IndexOutOfRangeException(
+                $"The txId {txId} does not exist in the index.");
+
+    /// <inheritdoc />
+    public async Task<BlockHash> GetContainedBlockHashByTxIdAsync(TxId txId) =>
+        await TryGetContainedBlockHashByIdAsync(txId)
+        ?? throw new IndexOutOfRangeException(
+            $"The txId {txId} does not exist in the index.");
+
+    /// <inheritdoc />
+    public abstract bool TryGetContainedBlockHashById(TxId txId, out BlockHash containedBlock);
+
+    /// <inheritdoc />
+    public abstract Task<BlockHash?> TryGetContainedBlockHashByIdAsync(TxId txId);
+
+    /// <inheritdoc />
+    void IBlockChainIndex.Index(
+        BlockDigest blockDigest, IEnumerable<ITransaction> txs, CancellationToken stoppingToken) =>
+        IndexImpl(blockDigest, txs, null, stoppingToken);
+
+    /// <inheritdoc />
+    async Task IBlockChainIndex.IndexAsync(
+        BlockDigest blockDigest, IEnumerable<ITransaction> txs, CancellationToken stoppingToken) =>
+        await IndexAsyncImpl(blockDigest, txs, null, stoppingToken);
+
+    void IBlockChainIndex.SynchronizeForever<T>(
+        BlockChain<T> chain, CancellationToken stoppingToken)
+    {
+        var newBlockEncountered = new AsyncAutoResetEvent();
+        chain.TipChanged += (_, _) => newBlockEncountered.Set();
+        while (true)
+        {
+            ((IBlockChainIndex)this).Synchronize<T>(chain.Store, stoppingToken);
+            newBlockEncountered.Wait(stoppingToken);
+            if (stoppingToken.IsCancellationRequested)
+            {
+                return;
+            }
+        }
+    }
+
+    async Task IBlockChainIndex.SynchronizeForeverAsync<T>(
+        BlockChain<T> chain, CancellationToken stoppingToken)
+    {
+        var newBlockEncountered = new AsyncAutoResetEvent();
+        chain.TipChanged += (_, _) => newBlockEncountered.Set();
```

Comment:
Since SynchronizeAsync is idempotent, I guess it's okay to just poll periodically and we don't need to wait for a new block through `BlockChain<T>` param.